### ISSUE TITLE
[arrow-parens] enforce parens for chained functions

### ISSUE THREAD

**Issue by Krustal** [eslint/eslint#7322]
**Environment**
1. ESLint Version: 3.7.1
2. Node Version: 5.9.1
3. npm Version: 3.7.3

**Parser:** default

**Config**

``` javascript
{
  "extends": "airbnb",
  "plugins": [],
  "env": {
    "browser": true,
    "mocha": true,
    "node": true,
    "es6": true,
  },
  "globals": {
    "sinon": false
  },
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true,
      "experimentalObjectRestSpread": true,
    }
  },
  # Custom Disabled Rules
  rules: {
    camelcase: 0,
    max-len: 0,
    no-console: ["error", { allow: ["info", "warn", "error"] }],
    no-param-reassign: 0,
    no-unused-expressions: 0, # Issues with expect
    no-unused-vars: ["error", { "varsIgnorePattern": "^_", "argsIgnorePattern": "^_", "caughtErrors": "none" }],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/*_test.js",
          "**/*_spec.js",
          "**/test/**",
          "**/*_story.js",
          "**/*_story.jsx",
          "**/server.js",
          "**/prod_server.js"
        ]
      }
    ]
  }
}
```

Pertinent change AirBnB makes to this rule:

``` javascript
'arrow-parens': ['error', 'as-needed', {
      requireForBlockBody: true,
    }],
```
1. What we did:
   ![screenshot 2016-10-07 11 57 06](https://cloud.githubusercontent.com/assets/442471/19202038/600b951e-8c85-11e6-88cb-160a28896d35.png)
2. What you would like to happen
   We would like there to be an option to enforce parens for arrow functions that return another arrow function expression. 
3. What actually happened
   We would like to omit parens for short arrow functions that return an expression, in the case of a arrow function passed to a collection method:

``` javascript
myCollection.map(val => val + 2);
```

But, enforce parens for arrow functions with block bodies:

``` javascript
const myFunc = (param) => {
  // some logic
}
```

However, when chaining arrow functions we would like to always require parens for the arguments.

``` javascript
const myFunc = (param) => (base) => {
  const mod = param + 2;
  return mod + base;
}
```

Because the outer function is returning an expression, as the inner function, it hits the 1st case, but the inner function has a block and so requires params, which would require you to change the above to:

``` javascript
const myFunc = param => (base) => {
  const mod = param + 2;
  return mod + base;
}
```

**Comment by not-an-aardvark** [comment]
Thanks for the feature request.

Personally, I don't think this option should be included in core. In this example:

```js
const myFunc = param => (base) => {
  const mod = param + 2;
  return mod + base;
}
```

`myFunc` is returning an expression, and it doesn't have a block body, so I would not expect `param` to be parenthesized. I don't think a special exception for arrow functions that return other arrow functions is a common enough use-case to merit an option in the core rule. (Of course, you are welcome to create a custom rule for your use-case.)

However, let's leave the issue open to see what the rest of the team thinks.

**Comment by Krustal** [comment]
Thanks for the response.

I think the currying syntax of chaining arrow functions is gaining tractions due to libraries like Redux (where I originally ran into this), for middleware and thunks, and React container components. 

I can see an argument for this being too specific but our team was finding it easier to visually parse this syntax when the function params were all defined consistently across the line and we didn't like no-parens for block functions.

**Comment by kaicataldo** [comment]
I'm with @not-an-aardvark on this one. I can see why you would want to enforce that style if you write a lot of code that looks like that, but it does seem a little too niche for core (I don't have much experience with React/Redux, though). Curious what others on the team think.

**Comment by SamMorrowDrums** [comment]
A sensible option coud be to enable enforcing paren-wrap for multi-line statements as per AirBnB style.

https://github.com/airbnb/javascript#arrows--paren-wrap

I ran into a recent error where somebody had written:

**Comment by platinumazure** [comment]
I haven't been tracking the conversation so apologies if I'm missing some context. That said, I want to remind everyone that even if we ultimately reject this for core, someone can always write a custom rule or a plugin for this, and that will fulfill the expressed need the same as a core rule would. ESLint was designed from the beginning to be extremely pluggable due to situations exactly like this one. So please don't be discouraged if we don't accept this proposal for ESLint core- you still have other options.

**Comment by kaicataldo** [comment]
@SamMorrowDrums I think that would be a change for http://eslint.org/docs/rules/arrow-body-style rather than the arrow-parens rule. Apologies if I'm missing something!

**Comment by SamMorrowDrums** [comment]
You're exactly right, I would love that option.

I was only proposing that the issue creator could achieve what they wanted
through this idea, as long as their functions bodies are multi-line.

I guess there is also the possibility for other restrictions, but
multi-line would be top of my list I think.

**Comment by not-an-aardvark** [comment]
It looks like there are several team members opposed to adding this option, so I'm going to go ahead and close this issue. Of course, you're always welcome to use this option on your own code if you implement it in a plugin.
